### ISSUE TITLE
no-unnecessary-mapfilter

### ISSUE THREAD

**Issue by hornta** [eslint/eslint#12269]
**Please describe what the rule should do:**
reports on unnecessary use of .filter and .map on arrays.
could be split into 2 rules, one for unnecessary use of .filter and one for unnecessary use of .map

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
const myArr = [1,2,3];
myArr.filter(x => true) // filter isn't filtering any items
```

```js
const myArr = [];
myArr.filter(x => 1)  // filter isn't filtering any items
```

```js
const myArr = [];
myArr.map(x => x) // map isn't mapping to something new, unless a shallow copy is the goal which in case one should go for .slice(0)
```

**Why should this rule be included in ESLint (instead of a plugin)?**
it's simple enough to be a part of eslint, maybe too simple to be a part of eslint? ❓ 

**Are you willing to submit a pull request to implement this rule?**
yes

**Comment by ljharb** [comment]
How often does this happen in practice?

(also, you can't statically know if the intention is to get a shallow copy, thus making `.slice()` the appropriate fix)

**Comment by hornta** [comment]
I don't think it happens very often but I've come across 2 instances of it in our code base but then the filter/map functions did some weird things with outside variables only to return true(for .filter)/itself(for .map) in the end. Maybe a .forEach would've been better.
Addition to https://eslint.org/docs/rules/array-callback-return perhaps?

**Comment by kaicataldo** [comment]
I’m not sure how this could be enforced without a lot of false negatives due to the limitation of ESLint not having type information. I understand the argument that catching some instances of an error is better than none, but I also think it’s a dangerous precedent to set because folks using it might not be aware of the cases it misses.

I think this is a better fit for a custom rule (actually, I wonder if the case of the callback returning a literal could be enforced with `no-restricted-syntax`?)

**Comment by silverwind** [comment]
I do see value in a rule that warns if the return value of non-mutating `Array.prototype` functions is not consumed.

I often see `.map` used in the wild where the code would ignore the return value in which case `.map` could be replaced with a `.forEach` for the same outcome (and probably better performance).

**Comment by kaicataldo** [comment]
@silverwind I like that idea, as long as we can find a way to do so without false negatives/positives. Seems like a nice rule to pair with [`array-callback-return`](https://eslint.org/docs/rules/array-callback-return).

**Comment by ljharb** [comment]
I'm pretty sure performance would be largely identical.

**Comment by silverwind** [comment]
@ljharb `forEach` seems slightly faster for me:

```` console
$ time node -e 'new Array(3e7).fill(true).forEach(v => {const x = v})'
0.50s user 0.09s system 96% cpu 0.612 total
$ time node -e 'new Array(3e7).fill(true).map(v => {const x = v})'
0.67s user 0.16s system 96% cpu 0.855 total
````

**Comment by ljharb** [comment]
Nothing is fast with new Array, even when filled. Try `Array.from({ length: 3e7 })`

**Comment by silverwind** [comment]
Comes out about the same with that `Array.from` for me.

Anyways, the new rule could be called something like `array-consume-return-value` and apply to the same methods like [array-callback-return](https://eslint.org/docs/rules/array-callback-return) minus `Array.prototype.sort` because that one mutates.

**Comment by ljharb** [comment]
(That rule is also missing flat and flatMap, ftr)

**Comment by kaicataldo** [comment]
@silverwind Would you like to make a new issue with a formal proposal?

**Comment by silverwind** [comment]
Yes. I will after checking if there if there are any more methods that could benefit, not strictly limited to Array.

**Comment by silverwind** [comment]
Proposal at https://github.com/eslint/eslint/issues/12980.
